activation/certifier.go

```diff
@@ -0,0 +1,369 @@
+package activation
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/spacemeshos/poet/shared"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/localsql"
+	"github.com/spacemeshos/go-spacemesh/sql/localsql/certifier"
+)
+
+type CertifierClientConfig struct {
+	// Base delay between retries, scaled with the number of retries.
+	RetryDelay time.Duration `mapstructure:"retry-delay"`
+	// Maximum time to wait between retries
+	MaxRetryDelay time.Duration `mapstructure:"max-retry-delay"`
+	// Maximum number of retries
+	MaxRetries int `mapstructure:"max-retries"`
+}
+
+type Base64Enc struct {
+	Inner []byte
+}
+
+func (e Base64Enc) String() string {
+	return base64.RawStdEncoding.EncodeToString(e.Inner)
+}
+
+// Set implements pflag.Value.Set.
+func (e *Base64Enc) Set(value string) error {
+	return e.UnmarshalText([]byte(value))
+}
+
+// Type implements pflag.Value.Type.
+func (Base64Enc) Type() string {
+	return "Base64Enc"
+}
+
+func (e *Base64Enc) UnmarshalText(text []byte) error {
+	b, err := base64.StdEncoding.DecodeString(string(text))
+	if err != nil {
+		return err
+	}
+	e.Inner = b
+	return nil
+}
+
+type CertifierConfig struct {
+	Client CertifierClientConfig `mapstructure:"client"`
+}
+
+func DefaultCertifierClientConfig() CertifierClientConfig {
+	return CertifierClientConfig{
+		RetryDelay:    1 * time.Second,
+		MaxRetryDelay: 30 * time.Second,
+		MaxRetries:    5,
+	}
+}
+
+func DefaultCertifierConfig() CertifierConfig {
+	return CertifierConfig{
+		Client: DefaultCertifierClientConfig(),
+	}
+}
+
+type ProofToCertify struct {
+	Nonce   uint32 `json:"nonce"`
+	Indices []byte `json:"indices"`
+	Pow     uint64 `json:"pow"`
+}
+
+type ProofToCertifyMetadata struct {
+	NodeId          []byte `json:"node_id"`
+	CommitmentAtxId []byte `json:"commitment_atx_id"`
+
+	Challenge []byte `json:"challenge"`
+	NumUnits  uint32 `json:"num_units"`
+}
+
+type CertifyRequest struct {
+	Proof    ProofToCertify         `json:"proof"`
+	Metadata ProofToCertifyMetadata `json:"metadata"`
+}
```

Comment:
This can be a single type that can be filled from the info that is in `nipost.Post`, the separation of Post into proof and metadata is arguably arbitrary and it would make (at least to me) a lot more sense if a PoST is self contained.

The separation was only because of the ATX wire type but that shouldn't use any of these types anyway 🙂 

```suggestion
type CertifyRequest struct {
	Nonce   uint32 `json:"nonce"`
	Indices []byte `json:"indices"`
	Pow     uint64 `json:"pow"`
	
	NodeId          []byte `json:"node_id"`
	CommitmentAtxId []byte `json:"commitment_atx_id"`

	Challenge []byte `json:"challenge"`
	NumUnits  uint32 `json:"num_units"`
}
```

Reply:
This structure is expected by the certifier service: https://github.com/spacemeshos/post-rs/blob/e42f504ed4b70ea6c5d28609d44bcd79096acf2c/certifier/src/certifier.rs#L17-L20

Reply:
But that's an implementation detail that shouldn't influence API design?

Reply:
Wdym? These types are encoded as JSON and sent in the HTTP POST request.

Reply:
I mean that the reason that `Post` and `PostMetadata` are separate objects is not because it makes sense from an (Certifier) API perspective but to safe on bytes in an ATX, where the metadata exists as a separate entity to deduplicate values that would be present twice in the ATX if they were part of the `Post` type again resulting from the fact that we used to re-use the same types for wire encoding and inside the domain of the codebase.

It's not a big issue if we at some point in the future re-do the API with OpenAPI anyway, I'm just arguing that if we look at it from a pure API perspective a PoST is a single object and data/metadata separation is arbitrary for the use in the Certifier.

Reply:
It makes sense from the certifier API perspective because all the logic there also expects division into PoST (the proof) and metadata (the data required to verify it).